### PR TITLE
LTD-4406 - populate report summaries part 2: Correct misspelt report summaries.

### PR DESCRIPTION
**Pull request by stuaxo**
Before: `python manage.py suggest_summaries`  could generate a CSV file of GoodOnApplications along with prefixes and subjects that match the report_summary field (after normalisation to remove numbers).

After:
Add a `--mappings` parameter to allow further matches for simple spelling mistakes.
mappings is a CSV of existing repost_summaries and their corrected versions.

After that matches are found as before (though for these corrected report_summaries no normalisation occurs).

Also
A bug was fixed in the prefix matching to use the normalised_report_summary instead of the plain report_summary, this gains us one more match when run against our data.

The migration itself is a straight copy of the previous data migration we did for this with just the filenames / migration name changed.

Lots of improvements in clarity in the integration tests for this suggested by @depsiatwal 

TODO
- [ ] Ensure CI passes here (except the sha checks)
- [ ] Ensure content is correct and commited
- [ ] Update sha
- [ ] Ensure CI pass

### CODE REVIEW

File: api/applications/tests/test_suggest_summaries.py [hunk 1]
```diff
@@ -41,6 +41,38 @@ def test_annotate_normalised_summary_annotation(self, report_summary, expected_n
 
         assert normalised_report_summary == expected_normalised_report_summary
 
+    @parameterized.expand(
+        [
+            ("arts and crafts", {}, "arts and crafts"),
+            ("art and crafts", {}, "art and crafts"),
+            ("arts and craft", {"arts and craft": "arts and crafts"}, "arts and crafts"),
```
Comment (hnryjmes) [inline api/applications/tests/test_suggest_summaries.py:48]: can you explain a bit more for me why these specific test cases are used and what they are checking?

Comment (stuaxo) [reply api/applications/tests/test_suggest_summaries.py]: This should be tightened up now.

File: api/applications/management/commands/suggest_summaries.py [hunk 2]
```diff
@@ -87,21 +126,18 @@ def handle(self, *args, **options):
         }
 
         with open(filename, "w") as csvfile:
```
Comment (depsiatwal) [inline api/applications/management/commands/suggest_summaries.py:128]: Method is fairly large and cumbersome possible to break into smaller chunks to make it more readable and testable. for example getting CSV headers ?

**Review by depsiatwal — COMMENTED** [review]
not sure I follow all of it but code looks ok 
small suggested and it appears we need more unit tests

File: api/applications/management/commands/suggest_summaries.py (hunk 2 above)
Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:128]: It's a bit of an integration test for the management command, we setup some data then look at the data on the other side.

Pushed with more context, in the form of comments and typing.

There isn't really much to do about splitting up the CSV header reading - it tries to check the data in the dumbest / most exact way possible: while the management command deals with CSV - the test creates a list of strings which is compared with the file content read with `.splitlines()` - header included.

I tweaked the naming a little to maybe make this clearer; so the final assert is:

```
...
actual_file_lines = csv_path.read_text().splitlines()

self.assertCountEqual(expected_csv_file_lines, actual_file_lines)
```

(To be fair find djangos "assertCountEqual a bit confusingly named, when it compares the *content* not the amount of lines - it's nice to use the instances might be different, but the content is the same - such as this case)

I also split out the code to write the map file, though this is a method just for this one test.

**Comment by stuaxo** [comment]
> not sure I follow all of it but code looks ok small suggested...
 - this feels pretty familiar to me :D

> it appears we need more unit tests

There's one line in one branch I should probably cover.

File: api/applications/management/commands/suggest_summaries.py (hunk 2 above)
Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:128]: @depsiatwal I broke things up further; I added testing of the output on stderr (GOA with report summaries we can't map) - since this also uses the same CSV format, I moved the CSV generation there.

Naming should now be tighter - "expected_" is only used as a prefix for variables that are asserted on, making some names shorter.

This should make `coverage` happier about the one line of one branch it didn't like.

File: api/applications/management/commands/suggest_summaries.py [hunk 3]
```diff
@@ -87,21 +126,18 @@ def handle(self, *args, **options):
...
+            has_written_unmappables_csv_header = False
             writer = csv.DictWriter(csvfile, fieldnames=csv_headers, quoting=csv.QUOTE_ALL)
             writer.writeheader()
+
+            stderr_writer = csv.DictWriter(self.stderr, fieldnames=csv_headers, quoting=csv.QUOTE_ALL)
```
Comment (currycoder) [inline api/applications/management/commands/suggest_summaries.py:133]: Help me understand the two CSV writers; one writes to the file, the other writes to stderr?  They both seem to write the same thing, correct?

File: api/applications/management/commands/suggest_summaries.py [hunk 4]
```diff
@@ -87,21 +126,18 @@ def handle(self, *args, **options):
         }
 
         with open(filename, "w") as csvfile:
+            has_written_unmappables_csv_header = False
```
Comment (currycoder) [inline api/applications/management/commands/suggest_summaries.py:129]: Why is this necessary?  What stops us from `stderr_writer.writeheader()` after that variable is defined below?

File: api/applications/migrations/0076_back_populate_product_report_summary_prefix_and_suffix.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            "suggested_prefix_id",
+            "suggested_subject",
+            "suggested_subject_id",
+        ]
+        reader = csv.DictReader(csvfile, fieldnames=fieldnames, quoting=csv.QUOTE_ALL)
+        next(reader, None)  # skip the headers
+        yield from reader
+
+
+class Migration(migrations.Migration):
```
[the remark this answers is not in the document]
Comment (stuaxo) [reply api/applications/migrations/0076_back_populate_product_report_summary_prefix_and_suffix.py:26]: Yep.

The only reason I hadn't done so, so far is because of the way the migrations tests were interacting - the test for 0075 was breaking the migration for 0076 when run in 0075.

A regression on this wouldn't have been visible if I switched this to the newest one, it's fixed now - but it made me reluctant to immediately remove it.

This is due to the way the migration tests reset the migrations initially: it first runs all the migrations, including those in the future: when I overrode the directory that is used to grab the CSVs from to the test, one it overrode the future 0076 one too - now fixed by changing the mechanism used for the override, so it's not overridden for the whole test.

[the remark this answers is not in the document]
Comment (stuaxo) [reply api/applications/migrations/0076_back_populate_product_report_summary_prefix_and_suffix.py:26]: Done

File: api/applications/management/commands/suggest_summaries.py (hunk 4 above)
Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:129]: Initially, we haven't written the header, so this is set to False.

We loop through the happy path, where every report_summary has a suggested_subject, in this case hitting the `continue` and looping.

Then, we have our first item with no suggested_subject, we hit the end part of the loop - since it is not set, we do that and then write the header.

Then we write the data to stderr.

On the next loop, we have written the header so we skip that part - then we write the row.

Still: now I look at this, all these negatives are confusing.

Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:129]: Simplified the control-flow here by sticking things in an if-else and removing the `continue` - as well as being more readable I wouldn't be surprised if coverage is happy now.

File: api/applications/management/commands/suggest_summaries.py (hunk 3 above)
Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:133]: Report summary is made up of either: 
`"{report_summary_subject}"` or `"{report_summary_prefix}{report_summary_subject}"`

Not having a subject is invalid, and an error.

In the PR that preceeded this, we logged these, here we send them to `stderr`, we also borrowed the same format as the CSV for convenience.

So:  
GOA with valid report_summaries, with a suggested subject and optionally a suggested prefix go the file.
GOA with invalid report_summaries containing no prefix go to stdout.

I could have added a commandline option to specify a file these can go to, add tests etc - in the end I tweaked what we had to output the right format, and grab the output (which I have to do anyway as these ultimately end up in excel).

Comment (stuaxo) [reply api/applications/management/commands/suggest_summaries.py:133]: ^ Adding some of this to comments in the code.

**Review by currycoder — APPROVED** [review]
Happy for the partial coverage to be skipped with a pragma nocover.  We discussed this offline and agreed that this management command should be short lived so absolute 100% coverage is not necessary.

**Comment by stuaxo** [comment]
Rebased and now project codecov isn't happy 🤬

**Comment by stuaxo** [comment]
codecov/project is so strict even with pragma nocov the answer is

image

**Comment by stuaxo** [comment]
> Happy for the partial coverage to be skipped with a pragma nocover. We discussed this offline and agreed that this management command should be short lived so absolute 100% coverage is not necessary.

Parts of this have a slightly higher density than I usually go for, so some comments just describe what this does as opposed to the usual more useful how / why type comments.

**Comment by stuaxo** [comment]
> looks good, appreciate the extra explanation in comments

Parts of this have a slightly higher density than I usually go for, so some comments just describe what this does as opposed to the usual more useful how / why type comments.
